helm: clustermesh: add a job for autoconfiguring coredns mcsapi config

MrFreezeex opened the pull request:
First commit move the existing mcsapi option into a subsection for regrouping all mcsapi configuration together.

Second commit add a new command to autoconfigure CoreDNS to configure CoreDNS with the required MCS-API configuration.

Third commit add the helm code to wire up the command for the previous commit in helm installation and recommend installing CoreDNS v1.12.2+ (will hopefully be included in Kubernetes 1.35+ by default) that now support natively MCS-API. 

```release-note
clustermesh: helm: move MCS-API helm config and add a job to autoconfigure CoreDNS for MCS-API for CoreDNS v1.12.2+
```

MrFreezeex committed:
helm: clustermesh: move mcsapi option to dedicated subsection

Create a new dedicated subsection for mcsapi and deprecated the old
option to prepare for more mcsapi related options grouped together.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

gandro reviewed (commented):
Thanks! Helm-wise this looks okay to me on a first glance, but I'm not a fan of us pulling in yet another external image. 

Have you investigated if this could be implemented in Go and make the job invoke a `clustermesh-apiserver` subcommand (or perhaps even run it a go-routine of one of the existing components)?

giorio94 commented on Documentation/operations/upgrade.rst:
```diff
@@ -288,6 +288,17 @@ communicating via the proxy must reconnect to re-establish connections.
+* The Helm option ``clustermesh.enableMCSAPISupport`` has been deprecated in favor of ``clustermesh.mcsapi.enabled``
+  and will be removed in Cilium 1.20.
+
 .. _1.18_upgrade_notes:
```
The old upgrade notes are typically removed when starting a new development cycle, preserving only the main section headers. 

/cc @aanm @joestringer I wonder if this task should be codified as part of the branch preparation to ensure consistency, rather than leaving it to the authors/reviewers.

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -3288,7 +3288,7 @@ clustermesh:
   # the remote endpoints of appropriately-annotated global services through ClusterMesh
   enableEndpointSliceSynchronization: false
-  # -- Enable Multi-Cluster Services API support
+  # -- Enable Multi-Cluster Services API support (deprecated; use clustermesh.mcsapi.enabled)
```
References to this option inside the documentation should be updated as well.

giorio94 commented on Documentation/network/clustermesh/mcsapi.rst:
```diff
@@ -43,60 +47,26 @@ To enable MCS-API support on an existing Cilium installation, run:
-MCS-API resources (ServiceImports). You can do so by running the following
-command on each cluster:
+By default Cilium will automatically configure and rollout CoreDNS for MCS-API support.
+You can also do it manually by setting the Helm value ``clustermesh.corednsMCSAPIAutoconfigure.enabled``
```
This option should likely be updated.

giorio94 commented on Documentation/network/clustermesh/mcsapi.rst:
```diff
@@ -43,60 +47,26 @@ To enable MCS-API support on an existing Cilium installation, run:
-You also need to make sure that CoreDNS is able to read and watch the relevant
-MCS-API resources (ServiceImports). You can do so by running the following
-command on each cluster:
+By default Cilium will automatically configure and rollout CoreDNS for MCS-API support.
```
I'm personally not a big fan of having this enabled by default, because of the wide heterogeneity of environments, which make it fairly difficult to predict in advance whether the automatic setup would work of not. Especially when dealing with cloud provider setups, which are often immutable, as well as in case users leverage GitOps solutions that would automatically revert any external changes.

MrFreezeex replied:
Hmm indeed so I thought about GitOps users but in most cases CoreDNS is still managed by kubeadm/what installs the cluster, some people might still do this to customize coredns via some GitOps tooling though.

For the cloud providers indeed I think you pretty much have to self manage CoreDNS/opting-out of the cloud provider doing this for you and thus should likely do this with their own tooling (meaning this auto configure won't be really useful for you).

So generally speaking I guess this would be mostly useful for users managing their clusters with kubeadm without disabling the coredns addon (or us with kind in dev environments). So indeed this seems good to have it disabled by default!

giorio94 commented on Documentation/operations/upgrade.rst:
```diff
@@ -288,6 +288,19 @@ communicating via the proxy must reconnect to re-establish connections.
+1.19 Upgrade Notes
+------------------
+
+* MCS-API CoreDNS configuration recommendation has been updated see :ref:`clustermesh_mcsapi_prereqs` for more details.
```
```suggestion
* MCS-API CoreDNS configuration recommendation has been updated. See :ref:`clustermesh_mcsapi_prereqs` for more details.
```

giorio94 commented on Documentation/network/clustermesh/mcsapi.rst:
```diff
@@ -43,60 +47,26 @@ To enable MCS-API support on an existing Cilium installation, run:
+      # Configure CoreDNS to support MCS-API
       kubectl get configmap -n kube-system coredns -o yaml | \
-         sed -E -e 's/^(\s*)kubernetes.*cluster\.local.*$/\1multicluster clusterset.local\n&/' | \
+         sed -e 's/cluster\.local/cluster.local clusterset.local/g' | \
```
Nit. I'd use `-E` in both cases for consistency, unless there's a strong reason not to.

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-coredns-mcsapi/job-role.yaml:
```diff
@@ -0,0 +1,24 @@
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
+  {{- with .Values.clustermesh.annotations }}
+  annotations:
```
These accessory artifacts should also have the `helm.sh/hook` annotation, so that they get automatically removed alongside the job.

giorio94 reviewed (commented):
Thanks! A couple of initial comments inline.

Generally speaking, I'd also prefer to avoid pulling in yet another image, and replace the bash script with a sub-command of one of the already existing images. That would also make the overall command more robust compared to bash scripting, and easier to extend in future, especially in case more advanced tasks need to be performed. Additionally, kubectl has technically a relatively strict compatibility policy, making it tricky to pick the correct version (even though the commands here are simple enough that real problems are unlikely). 

I'm not totally sure about the best image to use though. The `clustermesh-apiserver` seems the most likely candidate, even though it is not strictly necessary to run it to make MCS-API work. However, that would not be a significant problem as long as it is run as a separate job, which to me makes sense given its one-off nature.

MrFreezeex committed:
helm: docs: add job for MCS-API CoreDNS auto configuration

Starting in CoreDNS v1.12.2, CoreDNS natively supports MCS-API. This
commit update the existing documentation to reflect and recommend that.

This is also wiring a opt-in job in the helm installation to auto configure
CoreDNS that was added in the previous commit.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: add coredns autoconfiguration cmd in clustermesh-apiserver

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex commented on install/kubernetes/cilium/templates/clustermesh-coredns-mcsapi/job-clusterrole.yaml:
```diff
@@ -0,0 +1,19 @@
+rules:
+- apiGroups: [ "" ]
+  resources: [ "namespaces" ]
+  verbs: [ "get" ]
```
FYI this is apparently needed to setup the kubernetes client, I was getting a bunch of error of a client getting some kubernetes namespace without this.

giorio94 replied:
Could be worth adding a comment to the file itself, just for future reference.

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
```diff
@@ -0,0 +1,163 @@
+			}
+		},
+		PreRun: func(cmd *cobra.Command, args []string) {
+			// Overwrite the metrics namespace with the one specific for KVStoreMesh
```
Nit. Copy-paste leftover.

giorio94 commented on clustermesh-apiserver/cmd/root.go:
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/cilium/cilium/clustermesh-apiserver/clustermesh"
 	clustermeshdbg "github.com/cilium/cilium/clustermesh-apiserver/clustermesh-dbg"
 	"github.com/cilium/cilium/clustermesh-apiserver/common"
+	corednsMCSAPIAutoConfigure "github.com/cilium/cilium/clustermesh-apiserver/coredns-mcsapi-auto-configure"
```
Nit. How about `coredns-mcsapi-cfg`, or `coredns-mcsapi-autocfg` to shorten a bit the directory/package name? I'd also probably put mcsapi first (`mcsapi-coredns-cfg`), but again just a matter of preference.

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
Nit. I'd match the structure of the other clustermesh-apiserver commands, with a `root.go` file containing the actual command.

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
```diff
@@ -0,0 +1,163 @@
+		var corefile string
+		var found bool
+		if corefile, found = coreDNSConfigmap.Data["Corefile"]; !found {
+			err = fmt.Errorf("Corefile not found in ConfigMap %s/%s\n", config.CoreDNSNamespace, config.CoreDNSConfigMapName)
```
Nit. `\n` should not be here.

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
```diff
@@ -0,0 +1,163 @@
+	}
+
+	clusterDomainEscaped := strings.ReplaceAll(clusterDomain, ".", "\\.")
+	kubernetesMatchRegex := regexp.MustCompile(fmt.Sprintf(`(?m)^\s*kubernetes.*%s.*\{`, clusterDomainEscaped))
```
The regex is partially user-controlled, so it would be better to use `regexp.Compile` and check for errors, to prevent unexpected panics.

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
```diff
@@ -0,0 +1,163 @@
+	kubernetesMatchRegex := regexp.MustCompile(fmt.Sprintf(`(?m)^\s*kubernetes.*%s.*\{`, clusterDomainEscaped))
+	if !kubernetesMatchRegex.MatchString(corefile) {
+		err := fmt.Errorf("CoreDNS not configured with kubernetes plugin and the domain '%s'", clusterDomain)
+		logger.Error(err.Error())
```
Nit. Leaving logging to the caller seems slightly more idiomatic to me.

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
```diff
@@ -0,0 +1,163 @@
+		}
+		if corefile == "" {
+			return
+		}
```
Nit. Having a dedicated return value or a constant would make this case more explicit. Logging could also be performed here at the call site, again for the same reason.

MrFreezeex replied:
Now that logging is performed at the top level I think it's a bit more readable and since there is only one case like that it looks fine to me as is but if you feel that it should be improved I can definitely look into adding some kind of constant

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
```diff
@@ -0,0 +1,163 @@
+	return corefile, nil
+}
+
+func restartCoreDNS(ctx context.Context, logger *slog.Logger, client client.Clientset, config coreDNSConfig) error {
```
Nit. I would again defer logging to the caller.

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		logger.Error("Failed to patch CoreDNS Deployment", logfields.Error, err)
+		return err
+	}
```
Nit. You could use server-side apply (or just a plain update), instead of patch, which is a bit more convoluted:

```go
	_, err = client.AppsV1().Deployments(config.CoreDNSNamespace).Apply(ctx,
		applyappsv1.Deployment(config.CoreDNSDeploymentName, config.CoreDNSNamespace).
			WithAnnotations(map[string]string{
				"kubectl.kubernetes.io/restartedAt": time.Now().Format(time.RFC3339),
			}),
		metav1.ApplyOptions{FieldManager: "mcsapi-coredns-autocfg", Force: true})
```

MrFreezeex replied:
That looks great, thanks for the suggestion! :pray:

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd_test.go:
```diff
@@ -0,0 +1,206 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestUpdateCorefiles(t *testing.T) {
```
You could use a script test to validate the entire cell, instead of only the `updateCorefile` function.

giorio94 commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
```diff
@@ -0,0 +1,163 @@
+
+func updateCorefile(logger *slog.Logger, clusterDomain, clustersetDomain string, corefile string) (string, error) {
+	if strings.Contains(corefile, clustersetDomain) || strings.Contains(corefile, "multicluster") {
+		logger.Warn("CoreDNS might already have MCS-API configuration, skipping configuration")
```
Nit. This should be probably emitted at `Info` level, as always expected after the first run.

gandro replied:
I wonder if `level=warn` is the right choice here. What happens it the user does a Cilium upgrade via `helm upgrade` - will the job be re-triggered and it it always emit a warning? In that case, I think it might be better to leave this at `level=info`

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-coredns-mcsapi/job-role.yaml:
```diff
@@ -0,0 +1,25 @@
+  resourceNames: [ "{{ .Values.clustermesh.mcsapi.corednsAutoConfigure.coredns.configMapName }}" ]
+- apiGroups: [ "apps" ]
+  resources: [ "deployments" ]
+  verbs: [ "patch", "get" ]
```
I would add update as well, just to avoid surprises.

gandro commented on install/kubernetes/cilium/templates/clustermesh-coredns-mcsapi/job-clusterrole.yaml:
```diff
@@ -0,0 +1,19 @@
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
+  annotations:
+    "helm.sh/hook": post-install,post-upgrade
```
Is this Helm hook on the `ClusterRole`, `ClusteRoleBinding` and `ServiceAccount` intentional? It should only be necessary on the `Job` object I think?

giorio94 replied:
I'm the one to blame here actually. I felt it was more natural to have these associated resources removed automatically together with the job itself, but I don't have a strong opinion if you feel differently.

gandro replied:
I see. I'm fine with keeping them as is, I don't feel strongly. But we should be consistent (we have some other job with hooks), so might be worth updating those as well.

MrFreezeex replied:
I have removed those for consistency with the rest then!

gandro commented on install/kubernetes/cilium/templates/clustermesh-coredns-mcsapi/job-serviceaccount.yaml:
```diff
@@ -0,0 +1,20 @@
+  labels:
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
+
```
nit: Unnecessary empty line here

gandro reviewed (approved):
Thanks! I didn't look too much at the new Go code, but Helm looks good to me. One minor thing that I noticed

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-coredns-mcsapi/clusterrole.yaml:
```diff
@@ -0,0 +1,24 @@
+{{- if and (or .Values.clustermesh.mcsapi.enabled .Values.clustermesh.enableMCSAPISupport) .Values.clustermesh.mcsapi.corednsAutoConfigure.enabled }}
```
One tricky aspect here is that disabling MCS-API, or the CoreDNS autoconfiguration, or uninstalling Cilium would leave CoreDNS in a broken state, because it would still include the `multicluster` plugin, but the ClusterRole/ClusterRoleBinding would be removed. I'm not sure about the best approach though --- maybe applying the `helm.sh/resource-policy: keep` annotation to orphan it? Or moving the creation to the tool itself to avoid issues in case of re-installations? Having the tool do the revert on cleanup seems fairly complex/fragile to me instead, and it would risk incorrectly reverting user changes.

MrFreezeex replied:
Indeed very good catch, thanks a lot! I added the resource-policy annotation as your suggested as I think it's simpler than managing it in our code. This resource shouldn't really change that much I think but we could still manage it in code if we ever need to :man_shrugging:!

gandro commented on clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:
```diff
@@ -0,0 +1,163 @@
+		logger.Info("CoreDNS ConfigMap was updated successfully")
+
+		err = restartCoreDNS(ctx, logger, client, config)
+		if err != nil {
```
should this be `err == nil`?

giorio94 reviewed (commented):
Thanks! The new command makes sense to me. I've left a few more comments inline, mostly minor nits.

One additional slight weirdness is that the command is panicing if the kubeconfig is not specified, ~~or not correct~~ (it would be better providing a clearer error):

```
$ gor ./clustermesh-apiserver coredns-mcsapi-auto-configure

time=2025-07-15T10:24:45.030738639+02:00 level=info msg="Starting MCS-API CoreDNS auto configuration" module=coredns-mcsapi-auto-configure
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xc8 pc=0x24a134d]

goroutine 91 [running]:
k8s.io/client-go/kubernetes.(*Clientset).CoreV1(...)
        /home/marco/Documents/isovalent/cilium/vendor/k8s.io/client-go/kubernetes/clientset.go:326
github.com/cilium/cilium/clustermesh-apiserver/coredns-mcsapi-auto-configure.configureCoreDNS.func1({0x344b1d0, 0xc0009bfe30}, {0x7?, 0xfffffffffffffffc?})
        /home/marco/Documents/isovalent/cilium/clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go:67 +0x122
github.com/cilium/hive/job.(*jobOneShot).start(0xc000b135c0, {0x344b1d0, 0xc0009bfe30}, 0x0?, {0x3450460, 0xc000b13560}, {{{0xc000a81bc0, 0x1, 0x1}}, 0xc0009d3650, ...})
        /home/marco/Documents/isovalent/cilium/vendor/github.com/cilium/hive/job/oneshot.go:138 +0x4fd
created by github.com/cilium/hive/job.(*queuedJob).Start.func1 in goroutine 1
        /home/marco/Documents/isovalent/cilium/vendor/github.com/cilium/hive/job/job.go:126 +0x16f
```

Edit: you should check `client.IsEnabled()`.

giorio94 commented on clustermesh-apiserver/mcsapi-coredns-cfg/root.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+func configureCoreDNS(jg job.Group, logger *slog.Logger, shutdowner hive.Shutdowner, client client.Clientset, config coreDNSConfig) {
+	jg.Add(job.OneShot("clustermesh-nodemanager-notifier", func(ctx context.Context, _ cell.Health) (err error) {
```
Nit. The job name should be updated.

MrFreezeex replied:
Woops indeed :man_facepalming:

giorio94 commented on clustermesh-apiserver/mcsapi-coredns-cfg/root.go:
```diff
@@ -0,0 +1,172 @@
+
+		if !client.IsEnabled() {
+			err = fmt.Errorf("Kubernetes client is not enabled, cannot configure CoreDNS")
+			fmt.Println(err.Error())
```
Nit. Should use the logger.

giorio94 commented on clustermesh-apiserver/mcsapi-coredns-cfg/root.go:
```diff
@@ -0,0 +1,172 @@
+		} else {
+			logger.Info("CoreDNS is rolling out with the new configuration")
+		}
+		return
```
Nit. This return looks redundant (not sure if the linter would complain or not).

MrFreezeex replied:
Yep that's because of the named return error, AFAIK there has to be a return here for the code to compile unfortunately

giorio94 commented on clustermesh-apiserver/mcsapi-coredns-cfg/root.go:
```diff
@@ -0,0 +1,172 @@
+		ctx,
+		applyappsv1.Deployment(config.CoreDNSDeploymentName, config.CoreDNSNamespace).
+			WithAnnotations(map[string]string{
+				"kubectl.kubernetes.io/restartedAt": time.Now().Format(time.RFC3339),
```
Should we use a custom annotation here?

MrFreezeex replied:
Hmmm sure I am adding this maybe later this might be handy to check if we ran this before :thinking:

giorio94 commented on clustermesh-apiserver/mcsapi-coredns-cfg/root.go:
```diff
@@ -0,0 +1,172 @@
+	_, err = client.AppsV1().Deployments(config.CoreDNSNamespace).Apply(
+		ctx,
+		applyappsv1.Deployment(config.CoreDNSDeploymentName, config.CoreDNSNamespace).
+			WithAnnotations(map[string]string{
```
Just realized that I had forgotten a piece in my previous suggestion:

```suggestion
		applyappsv1.Deployment(config.CoreDNSDeploymentName, config.CoreDNSNamespace).
			WithSpec(applyappsv1.DeploymentSpec().WithTemplate(
				v1.PodTemplateSpec().WithAnnotations(map[string]string{
```

MrFreezeex replied:
Ah yep indeed :sweat_smile:

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-coredns-mcsapi/clusterrole.yaml:
```diff
@@ -0,0 +1,25 @@
+    {{- toYaml . | nindent 4 }}
+    {{- end }}
+  annotations:
+    helm.sh/resource-policy: keep
```
Nit. I would add a comment to explain why this is here (and similarly for the other occurrence).

giorio94 reviewed (commented):
Looks good to me, thanks! I'll take a final look once the PR is marked ready for review.

MrFreezeex committed:
clustermesh: test mcsapi-coredns-cfg with hive script

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex commented on clustermesh-apiserver/mcsapi-coredns-cfg/testdata/configure.txtar:
```diff
@@ -0,0 +1,216 @@
+# Test a normal CoreDNS deployment being configured
```
I wasn't super sure how to do the error cases with hive script because there is no "outcome" we are just expecting the same initial state to stay... So I just ended up doing only the valid case and letting the unit test as is in this current version.

I am guessing that to test the error cases with hive scrip we would have to sleep for like 1 second and checking that we haven't changed anything but doesn't feel right compared to the current unit test :man_shrugging:.

giorio94 replied:
Yeah, I agree with your approach. This is mostly to test "E2E" the happy path, while unit tests are better suited for testing the error cases in this context.

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-coredns-mcsapi/job.yaml:
```diff
@@ -0,0 +1,83 @@
+      volumes:
+      {{- toYaml . | nindent 6 }}
+      {{- end }}
+      affinity:
```
I guess this (`affinity:`)  should be moved one line below (i.e., inside the `with` statement).

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-coredns-mcsapi/job.yaml:
```diff
@@ -0,0 +1,83 @@
+      tolerations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      serviceAccount: {{ .Values.serviceAccounts.corednsMCSAPI.name | quote }}
```
Nit. This is a deprecated alias, and could be omitted.

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -3327,6 +3333,49 @@ clustermesh:
     # -- Enable Multi-Cluster Services API support
     enabled: false
+    corednsAutoConfigure:
+      # -- Enable auto-configuration of CoreDNS for Multi-Cluster Services API.
+      #    CoreDNS MUST be at least in version v1.12.2 to run this.
+      enabled: false
```
Could it be worthy validating the version as part of the tool logic?

MrFreezeex replied:
I was a bit worried about some custom CoreDNS build that we wouldn't be able to check very well but I ended up adding a check that should be good enough for most image. And if it can't parse the version we only issue a warning which should be better than nothing IMO

giorio94 replied:
Makes sense to me, thanks!

giorio94 commented on clustermesh-apiserver/mcsapi-coredns-cfg/root.go:
```diff
@@ -0,0 +1,175 @@
+			logger.Info("CoreDNS might already have MCS-API configuration, skipping configuration")
+			return
+		}
+		coreDNSConfigmap.Data["Corefile"] = corefile
```
I wonder if it could make sense to save the previous version under another key (e.g., `Corefile.bak`), just to make it easy to restore it in case of necessity.

giorio94 commented on clustermesh-apiserver/mcsapi-coredns-cfg/script_test.go:
```diff
@@ -0,0 +1,103 @@
+			cell.Provide(func(db *statedb.DB) (kvstore.Client, uhive.ScriptCmdsOut) {
+				client := kvstore.NewInMemoryClient(db, "__local__")
+				return client, uhive.NewScriptCmds(kvstore.Commands(client))
+			}),
```
Nit. This is redundant in this context, and it could be dropped for simplicity.

giorio94 commented on clustermesh-apiserver/mcsapi-coredns-cfg/testdata/configure.txtar:
```diff
@@ -0,0 +1,216 @@
+# Start the hive to launch the CoreDNS configuration job
+hive start
+
+# Check if configmap are updated
```
Nit. `s/are/is/`.

giorio94 reviewed (approved):
Thanks! A couple of final comments inline.

tommyp1ckles reviewed (approved):
Not as familiar with clustermesh but k8s helm changes seem reasonable to me.  Approving for @cilium/sig-k8s.
